### ISSUE TITLE
Use full `header_footer: true` Asciidoctor output

### ISSUE THREAD

**Issue by mattpalermo** [asciidoctor/asciidoctor-browser-extension#217]
Many small problems arise (#86, #122, #208, and probably many others) from using the embedded Asciidoctor output instead of the full page output produced with `header_footer: true`. I appreciate the difficulties associated with this so I experimented with a proof of concept to figure out some useful ways to implement it: https://github.com/mattpalermo/asciidoctor-browser-renderer/blob/v0.1.0/src/main.js

As shown in the `main.js` linked above, the `DomParser` API can be used to parse the Asciidoctor output into HTML. This gives us some powerful ways to pick and choose what we want to copy into the current browser document. In the demonstration, I could simply use `document.documentElement.innerHTML = doc.documentElement.innerHTML;` to copy the entire document. Note that `documentElement.outerHTML` is read only :(. But, since this extension does some extra stuff in the `<head>` of the document, it is possible to handle things in more carefully.

`document.body = doc.body` works great and gives us the body classes required to solve issue #86. `document.head` is unfortunately read only, but we can do `document.head.innerHTML = doc.head.innerHTML`. More careful handling of the copy can be achieved through any number of the typical DOM methods.

One caveate of the `DomParser` is that `<script>`s are not executed. This can be solved, at least in a crude way, by replacing each `<script>` element with a copy of itself as shown in `main.js` linked above.

**Comment by Mogztter** [comment]
Thanks for your great work and detailed explanation! :+1: 
Now we know that we can replace the full page :beers: 

I have two pending pull requests but then I want to explore this idea :smiley: 

With this solution, I think it will be easy to integrate the Reveal.js converter: https://github.com/asciidoctor/asciidoctor-reveal.js

> document.body = doc.body works great and gives us the body classes required to solve issue #86. document.head is unfortunately read only, but we can do document.head.innerHTML = doc.head.innerHTML. More careful handling of the copy can be achieved through any number of the typical DOM methods.

Good to know :wink: 

> One caveate of the DomParser is that <script>s are not executed. This can be solved, at least in a crude way, by replacing each <script> element with a copy of itself as shown in main.js linked above.

`innerHTML` will not execute `<script>` but if we manually add `<script>` to the DOM it should work (ie. with `document.body.appendChild`).

But to be honest I'm considering dropping this "feature" for inline `<script>` defined in the `<body>` so that's not a big deal.

@mojavelinux What do you think about this proposal ?

**Comment by mattpalermo** [comment]
> I have two pending pull requests but then I want to explore this idea

Awesome! I can't wait.

> With this solution, I think it will be easy to integrate the Reveal.js converter

Maybe it'll be as simple as swapping out the `processor` for anything that produces a HTML document. At some point someone might like `asciidoctor-latex` as well. One question is: How would the add-in know when to use `Asciidoctor` and when to use one of these extensions?

> `innerHTML` will not execute `<script>`

Yes, `innerHTML` and `DomParser` seem to work very similarly for scripts at least. Probably for good reason.

> But to be honest I'm considering dropping this "feature" for inline `<script>` defined in the `<body>` so that's not a big deal.

Regardless of this feature, Asciidoctor will still output it's own scripts that it expects to be run. The most relevant one to me is the MathJax script included at the end of the `<body>` if `:stem:` is specified. Similarly, there are other scripts output by Asciidoctor (like highlight.js) or other processors (like asciidoctor-reveal) that need to be run in order to have the document display properly.

**Comment by Mogztter** [comment]
> Regardless of this feature, Asciidoctor will still output it's own scripts that it expects to be run. The most relevant one to me is the MathJax script included at the end of the <body> if :stem: is specified. Similarly, there are other scripts output by Asciidoctor (like highlight.js) or other processors (like asciidoctor-reveal) that need to be run in order to have the document display properly.

You are right... I don't want to use `eval` so I guess our only alternative is to do what you are doing. For the reference I'm already doing something similar for `<iframe>`: 

https://github.com/asciidoctor/asciidoctor-browser-extension/blob/f4961737a468ff1d485c5a8db7db29872700ad1e/app/js/renderer.js#L349-L353

I think JQuery is using `eval` because it's faster than manipulating the DOM ?

**Comment by mattpalermo** [comment]
It's a tricky problem to solve that I don't know much about. I'm not sure if there are also asynchronous issues to deal with. I see your `<iframe>` solution looks very similar! You even have the mention of scripts in the comment. I forgot about iframes.

**Comment by mojavelinux** [comment]
I understand what you're aiming for, but I don't think using the standalone document output is the way to go. That severely limits the control the browser extension has. What we should do instead is simply close the gap between integrations that the browser extension and what core provides.

The sidebar toc, in particular, is such a straightforward problem to solve. But to go further, the browser extension could display the toc in a much more integrated way than the standalone document does...and thus has the freedom to innovate. If all the extension did was display the standalone document, there would be no room for innovation or custom elements. And no custom themes. And no custom scripts. And a lot more output to wade through.

**Comment by mojavelinux** [comment]
Keep in mind, though, that if the embeddable document is too limited, we can address that upstream in core. A good example is the document header (which shows the document title, author, etc). Originally, the embeddable document had no header at all. But now that it does, maybe it should be more complete. If it's warranted, we should have that discussion.

**Comment by mojavelinux** [comment]
> At some point someone might like asciidoctor-latex as well. One question is: How would the add-in know when to use Asciidoctor and when to use one of these extensions?

It's reasonable to assume we can provide an option for which converter to use, so long as the converter works with Asciidoctor.js. We'd still ask the converter for the embeddable HTML, but we may also ask it for other things as well.

**Comment by mattpalermo** [comment]
You have a good point. It is a bit of a hack that sacrafices flexibility for quick results. It will take more work/thought but will be worth it for the final result. Especially since it will guide the development of the embeddable output to something that is undoubtably more powerful and flexible then the static document output.

**Comment by mojavelinux** [comment]
> undoubtably more powerful and flexible then the static document output.

Exactly. That's an unstated goal of this extension, IMO. In fact, I think its worth explicitly stating as a goal in the README. It's more than just a preview. It's about a nice integration (the refresh feature being a primary example of that).

**Comment by ggrossetie** [comment]
Fixed in #704
